File: hash/js/rosetta-squint-hash/src/multiHash.ts
/**
 * ImageMultiHash — container for crop_resistant_hash output.
 *
 * Each instance wraps a list of Hash objects (one per image segment).
 * Distance semantics differ from Hash: subtract() returns a FLOAT (not int),
 * because the matching algorithm weighs matches vs. unmatched segments.
 */

import { Hash, ImageHashError } from "./hash.js";
import { hexToHash } from "./hex.js";

export class ImageMultiHash {
  /**
   * Throws `ImageHashError("ShapeMismatch", ...)` if `segmentHashes` is empty —
   * matches Java's existing guard so all 5 ports reject empty multi-hashes at
   * construction rather than failing later inside `hashDiff` (which would
   * dereference `segmentHashes[0]`).
   */
  constructor(public readonly segmentHashes: Hash[]) {
    if (segmentHashes.length === 0) {
      throw new ImageHashError("ShapeMismatch", "ImageMultiHash requires at least one segment hash");
    }
  }

  /**
   * hash_diff: returns (matches, sumDistance) matching Python's hash_diff.
   *
   * For each segment in `this`, find the closest segment in `other` (by Hamming
   * distance). If that distance <= hammingCutoff, count it as a match.
   *
   * Default hammingCutoff = bitErrorRate * bitCount (bitErrorRate defaults to 0.25).
   */
  hashDiff(
    other: ImageMultiHash,
    hammingCutoff?: number,
    bitErrorRate?: number,
  ): { matches: number; sumDistance: number } {
    // Resolve cutoff
    let cutoff = hammingCutoff;
    if (cutoff === undefined) {
      const ber = bitErrorRate !== undefined ? bitErrorRate : 0.25;
      cutoff = this.segmentHashes[0].bitCount() * ber;
    }

    let matches = 0;
    let sumDistance = 0;

    for (const seg of this.segmentHashes) {
      let lowestDistance = Infinity;
      for (const otherSeg of other.segmentHashes) {
        const d = seg.subtract(otherSeg);
        if (d < lowestDistance) lowestDistance = d;
      }
      if (lowestDistance <= cutoff) {
        matches++;
        sumDistance += lowestDistance;
      }
    }

    return { matches, sumDistance };
  }

  /**
   * subtract: returns a FLOAT distance (matches Python's __sub__).
   *
   * Higher value = more different.  Returns maxDifference (= this.segmentHashes.length)
   * if no segments match.
   */
  subtract(
    other: ImageMultiHash,
    hammingCutoff?: number,
    bitErrorRate?: number,
  ): number {
    const { matches, sumDistance } = this.hashDiff(other, hammingCutoff, bitErrorRate);
    const maxDifference = this.segmentHashes.length;
    if (matches === 0) return maxDifference;
    const maxDistance = matches * this.segmentHashes[0].bitCount();
    const tieBreaker = 0 - sumDistance / maxDistance;
    const matchScore = matches + tieBreaker;
    return maxDifference - matchScore;
  }

  /**
   * matches: returns true if at least regionCutoff segments match.
   */
  matches(
    other: ImageMultiHash,
    regionCutoff = 1,
    hammingCutoff?: number,
    bitErrorRate?: number,
  ): boolean {
    return this.hashDiff(other, hammingCutoff, bitErrorRate).matches >= regionCutoff;
  }

  /**
   * bestMatch: returns the ImageMultiHash in `others` with the smallest subtract() distance.
   */
  bestMatch(
    others: ImageMultiHash[],
    hammingCutoff?: number,
    bitErrorRate?: number,
  ): ImageMultiHash {
    let best = others[0];
    let bestDist = this.subtract(best, hammingCutoff, bitErrorRate);
    for (let i = 1; i < others.length; i++) {
      const d = this.subtract(others[i], hammingCutoff, bitErrorRate);
      if (d < bestDist) {
        bestDist = d;
        best = others[i];
      }
    }
    return best;
  }

  /** Comma-separated hex of each segment hash, matching Python str(multihash). */
  toString(): string {
    return this.segmentHashes.map(h => h.toString()).join(",");
  }
}

/**
 * Parse a comma-separated hex string (as produced by ImageMultiHash.toString())
 * back into an ImageMultiHash.
 */
export function hexToMultiHash(s: string): ImageMultiHash {
  return new ImageMultiHash(s.split(",").map(hexToHash));
}

File: hash/rust/rosetta-squint-hash/src/multihash.rs
//! ImageMultiHash: variable-length list of Hash objects, output of crop_resistant_hash.
//!
//! Distance is FLOAT (not int) — matches Python `imagehash.ImageMultiHash.__sub__`.

use std::fmt;

use crate::{hex_to_hash, Hash, ImageHashError};

/// A collection of per-segment hashes produced by `crop_resistant_hash`.
///
/// Distance (`subtract`) returns a `f64` — not an integer — because it sums
/// best-match Hamming distances and applies a penalty for unmatched segments.
#[derive(Debug, Clone)]
pub struct ImageMultiHash {
    pub segment_hashes: Vec<Hash>,
}

impl ImageMultiHash {
    /// Returns the float distance between two `ImageMultiHash` objects.
    ///
    /// Matches Python `ImageMultiHash.__sub__`:
    /// ```python
    /// matches, sum_distance = self.hash_diff(other)
    /// max_difference = len(self.segment_hashes)
    /// if matches == 0: return max_difference
    /// max_distance = matches * len(self.segment_hashes[0])
    /// tie_breaker = -(sum_distance / max_distance)
    /// match_score = matches + tie_breaker
    /// return max_difference - match_score
    /// ```
    pub fn subtract(&self, other: &Self) -> f64 {
        self.subtract_with(other, None, None)
    }

    pub fn subtract_with(
        &self,
        other: &Self,
        hamming_cutoff: Option<f64>,
        bit_error_rate: Option<f64>,
    ) -> f64 {
        let (matches, sum_distance) = self.hash_diff(other, hamming_cutoff, bit_error_rate);
        let max_difference = self.segment_hashes.len() as f64;
        if matches == 0 {
            return max_difference;
        }
        let max_distance = matches as f64 * self.segment_hashes[0].bit_count() as f64;
        let tie_breaker = -(sum_distance as f64 / max_distance);
        let match_score = matches as f64 + tie_breaker;
        max_difference - match_score
    }

    /// Returns `(matches, sum_of_hamming_distances)` for segments within cutoff.
    ///
    /// Matches Python `ImageMultiHash.hash_diff`.
    pub fn hash_diff(
        &self,
        other: &Self,
        hamming_cutoff: Option<f64>,
        bit_error_rate: Option<f64>,
    ) -> (usize, usize) {
        if self.segment_hashes.is_empty() {
            return (0, 0);
        }
        let cutoff = match hamming_cutoff {
            Some(c) => c,
            None => {
                let ber = bit_error_rate.unwrap_or(0.25);
                self.segment_hashes[0].bit_count() as f64 * ber
            }
        };

        let mut total_matches = 0usize;
        let mut total_distance = 0usize;

        for sh in &self.segment_hashes {
            let lowest = other
                .segment_hashes
                .iter()
                .filter_map(|oh| sh.subtract(oh).ok())
                .min()
                .unwrap_or(usize::MAX);
            if (lowest as f64) <= cutoff {
                total_matches += 1;
                total_distance += lowest;
            }
        }
        (total_matches, total_distance)
    }

    /// True if at least `region_cutoff` segments match within the hamming cutoff.
    pub fn matches(
        &self,
        other: &Self,
        region_cutoff: usize,
        hamming_cutoff: Option<f64>,
        bit_error_rate: Option<f64>,
    ) -> bool {
        let (m, _) = self.hash_diff(other, hamming_cutoff, bit_error_rate);
        m >= region_cutoff
    }

    /// Returns the `ImageMultiHash` in `others` that minimizes `self.subtract(other)`.
    pub fn best_match<'a>(&self, others: &'a [Self]) -> Option<&'a Self> {
        others.iter().min_by(|a, b| {
            self.subtract(a)
                .partial_cmp(&self.subtract(b))
                .unwrap_or(std::cmp::Ordering::Equal)
        })
    }

    /// Comma-joined hex of each segment hash (matches Python `str(multihash)`).
    pub fn to_hex(&self) -> String {
        self.segment_hashes
            .iter()
            .map(|h| h.to_hex())
            .collect::<Vec<_>>()
            .join(",")
    }
}

impl fmt::Display for ImageMultiHash {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.to_hex())
    }
}

/// Parse a comma-separated multi-hash hex string into an `ImageMultiHash`.
///
/// Matches Python `imagehash.hex_to_multihash`.
pub fn hex_to_multihash(s: &str) -> Result<ImageMultiHash, ImageHashError> {
    let hashes: Result<Vec<Hash>, ImageHashError> =
        s.split(',').map(|part| hex_to_hash(part)).collect();
    Ok(ImageMultiHash { segment_hashes: hashes? })
}
